### PR TITLE
Implements ones_like for Sparse COO tensors

### PR DESCRIPTION
**Pull request by nikitaved**
This stack tries to eliminate `sparse_mask_helper` and improve `sparse_mask`.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #91714
* #91712
* #91704
* #92248
* #92242
* __->__ #91703

cc @pearu @cpuhrsch @amjames @bhosmer

### CODE REVIEW

File: aten/src/ATen/native/TensorFactories.cpp [hunk 1]
```diff
@@ -688,6 +688,29 @@ Tensor ones_like(
...
+    auto result = self.to(
+        dtype,
+        layout,
+        device,
+        pin_memory,
+        /*non_blocking=*/false,
+        /*copy=*/true,
+        optional_memory_format);
+
+    result.values().fill_(1.);
```
Comment (pearu) [inline aten/src/ATen/native/TensorFactories.cpp:710]: Hmm, we cannot do this for several reasons:
- recall, the layout is an implementation detail and this change violates invariant `op(sparse.to_dense()) == op(sparse)`
- `ones_like` definition is "`input. torch.ones_like(input)` is equivalent to `torch.ones(input.size(), dtype=input.dtype, layout=input.layout, device=input.device)`". We don't support `torch.ones(..., layout=torch.sparse_coo)` atm.

**Review by pearu — COMMENTED** [review]
I am not sure that this PR can be approved as it is.

We should first define `torch.ones(..., layout=torch.sparse_coo)` and then update the PR accordingly.

File: aten/src/ATen/native/TensorFactories.cpp (hunk 1 above)
Comment (pearu) [reply aten/src/ATen/native/TensorFactories.cpp:710]: Re 1: it is a problem as it says.

Re 2: your disagreement is with the `ones_like` documentation then.

We cannot use compressed format behavior as a first-hand reference. Some implemented features were introduced before we agreed on the semantics of the sparse tensors.

What is the reason for implementing `torch.ones_like` support for sparse tensors? Perhaps the code that currently uses `ones_like` needs to be updated...

Comment (nikitaved) [reply aten/src/ATen/native/TensorFactories.cpp:710]: Re 1. It is a factory method. What is the point in implementing `ones_like` that materializes all inputs and does not inherent the pattern?
Re 2. I have no problem with that. But somehow `empty_like` follows this suite, and I think it was approved to he like that. Why is this method any different?

Comment (cpuhrsch) [reply aten/src/ATen/native/TensorFactories.cpp:710]: @nikitaved On the intersection, you mean [the following snippet](https://github.com/pytorch/pytorch/pull/91704/files#diff-4adbd88239afcd60e8198aab65d4f5e43b62314e34b80551e997a1ea503adea5R1386-R1387) as an example?

```
grad.coalesce().mul(at::ones_like(result.coalesce()));
```

Hm, we could extend [`sparse_mask`](https://pytorch.org/docs/master/generated/torch.Tensor.sparse_mask.html?highlight=sparse_mask) to work with sparse inputs. So then we could have `grad.sparse_mask(result)`?

Comment (pearu) [reply aten/src/ATen/native/TensorFactories.cpp:710]: Re 1: `ones_like` is a factory method is a fair point.

Nevertheless, the suggested change contradicts how `torch.ones_like` is defined in docs (that seems a reasonable starting point for implementing features for sparse tensors, I hope you agree):
1. its result tensor is filled with ones;
2. there is an equivalence relation with `torch.ones`

The above makes total sense and has no ambiguity when considering strided tensors.

When considering sparse tensors, the definitions above can be understood in many ways. For instance, "a sparse tensor filled with ones" can either be understood as `(sparse.values() == 1).all()` (that this PR suggests) or `(sparse.to_dense() == 1).all()`.
Currently, we use sparse tensor semantics where the equality of tensors is the element-wise equality independent of the layout, and the sparse storage is considered as a memory optimization. Hence, with the sparse tensor semantics,  "a sparse tensor filled with ones" should be understood as `(tensor.to_dense() == 1).all()`.

Re `torch.empty_like`: while it has many similarities to `torch.ones_like` (e.g. it is a factory function, and has formally the same definition), it has no tensor semantics. In fact, `torch.empty` has properties that make sense only for strided tensors in the sense that "empty" strided tensor is a valid tensor, just its value content is undefined for efficiency. An "empty" sparse tensor (recall, a sparse tensor is a pair of strided tensors: `indices` and `values`, roughly speaking) can be a valid tensor only when the number of specified elements is zero, nse==0, (otherwise one would also need to specify a valid index structure but `torch.empty` API does not have methods for it). However, keeping in mind that `empty_like` has no tensor semantics, we can define `torch.empty_like` for sparse tensors so that the result will have the same index structure as the input, that is, nse>0 is possible.
To sum up, `ones_like` and `empty_like` methods have differences in having tensor semantics defined and not having tensor semantics defined, respectively. The latter gives a little more freedom when defining the methods for sparse tensors.

All the options to resolve this PR that I can think of, are painful (redefine sparse semantics, introduce tensor semantics-free `ones` method, etc), and hence my question:

> What is the reason for implementing `torch.ones_like` support for sparse tensors? ...

Comment (nikitaved) [reply aten/src/ATen/native/TensorFactories.cpp:710]: @pearu , I agree, but at the same time the documentation is clearly written for dense tensors, and sparse tensors play second fiddle. In fact, when it comes to sparse, I almost never read the documentation and jump straight into a terminal to see whether it is even supported, so it does not facilitate that much trust as it should. The name `*_like` is too powerful to assume borrowing just shape, device, dtype. It clearly suggests inheriting the structure and I doubt there is any other way to implemented `ones_like`. Sparse is getting evolved past just memory optimization, and deserves its place along side dense tensors, given how the coverage is getting improved, the autograd support, the real usage in large attention models. I think all this could be resolved with just defining what `like` even means in these `*_like` methods.

@cpuhrsch , yes, similar to that. Shall we extend it indeed? Otherwise I could go with a helper function that we only need in one file, luckily.

Comment (nikitaved) [reply aten/src/ATen/native/TensorFactories.cpp:710]: I will probably abandon this PR, and just use `mul` directly which is smart enough to handle view tensors so as to avoid any redundant memory allocations.

Comment (cpuhrsch) [reply aten/src/ATen/native/TensorFactories.cpp:710]: @nikitaved - do you still plan on abandoning this PR?

Comment (nikitaved) [reply aten/src/ATen/native/TensorFactories.cpp:710]: Yes, sparse_mask with sparse support is more performant.
